src/autograd.cpp

```diff
@@ -18,6 +19,23 @@
   return (void *)(ret);                                            \
 }
 
+namespace {
+
+void execute_backward_task(std::function<void>&& backward) {
+  auto const policy = (
+    // If we are in the main thread, then spawn another thread to carry out
+    // the `backward` task asynchronously.
+    // If we are already outside the main thread), then run `backward`
+    // synchronously in the current thread.
+    std::this_thead::get_id() == main_thread_id() ?
+      std::launch::async : std::launch::deferred
+  );
+
+  std::async(policy, std::move(backward)).get();
+}
```

Comment:
I'm hazarding a guess that we won't really need any of the task deque data structures (as indicated by the numerous red rectangles in the rest of this PR), and all we need is simply this... but I could be completely wrong :sweat_smile:

Reply:
I think at least one event loop will be needed because the R function that's called at some point by `lantern_autograd_backward` must be somehow scheduled to be executed in the main thread.